### PR TITLE
[cloud-provider-openstack] Add param etcdDiskSizeGb and migrate from openstack_blockstorage_volume_v2 to openstack_blockstorage_volume_v3

### PR DESCRIPTION
**Pull request by AleksZimin**
## Description
Added the `etcdDiskSizeGb` parameter for OpenStack and migration from `openstack_blockstorage_volume_v2` to `openstack_blockstorage_volume_v3`.

Yandex Cloud, GCP, and Azure `ClusterConfiguration` added in https://github.com/deckhouse/deckhouse/pull/4720
The `etcdDiskSizeGb` was not added to vSphere because there is no ability to resize `vsphere_virtual_disk` in `terraform-provider-vsphere`. See https://github.com/hashicorp/terraform-provider-vsphere/issues/851

## Why do we need it, and what problem does it solve?

The etcd disk size is now hardcoded. We've added a parameter to allow the user to choose the disk size, which solves the problem of the inflexible disk size in the previous version.
Partially closes issue https://github.com/deckhouse/deckhouse/issues/3348

We also migrated from `openstack_blockstorage_volume_v2` resource to `openstack_blockstorage_volume_v3`, as the former resource did not support disk resizing.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: candi
type: chore
summary: Added the `etcdDiskSizeGb` parameter for OpenStack and migration from `openstack_blockstorage_volume_v2` to `openstack_blockstorage_volume_v3`.
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AleksZimin. I've started the workflow for label 'e2e/run/openstack'...

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+const TerraformStateNamespace = "d8-system"
+const OpenstackV2ResourceType = "openstack_blockstorage_volume_v2"
+const OpenstackV3ResourceType = "openstack_blockstorage_volume_v3"
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnStartup: &go_hook.OrderedConfig{Order: 10},
+}, dependency.WithExternalDependencies(openstackTerraformStateMigration))
+
+func openstackTerraformStateMigration(input *go_hook.HookInput, dc dependency.Container) error {
+	input.LogEntry.Infof("Starting Terraform migrate hook")
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:39]: this is unnecessary

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	terraformStateLabels := map[string]string{
+		"node.deckhouse.io/terraform-state": "",
+	}
+
+	terraformStateSecrets, err := kubeCl.CoreV1().
+		Secrets(TerraformStateNamespace).
+		List(context.TODO(), metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(terraformStateLabels))})
+	if err != nil {
+		return fmt.Errorf("failed to get Terraform state secrets in namespace %s with labels %s. The migration process has been aborted", TerraformStateNamespace, terraformStateLabels)
+	}
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:54]: List is too expencive option, you should use get and check errors something like this -
https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsNotFound:~:text=func%20IsNotFound-,%C2%B6,-func%20IsNotFound(err

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+	if terraformStateSecrets.Items == nil {
+		input.LogEntry.Infof("Terraform state not found. Migration is not needed.")
+		return nil
+	}
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:59]: simply check https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsNotFound:~:text=func%20IsNotFound-,%C2%B6,-func%20IsNotFound(err

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+	for _, secret := range terraformStateSecrets.Items {
+		input.LogEntry.Infof("Proceeding with Secret/%s/%s", TerraformStateNamespace, secret.ObjectMeta.Name)
+		backupSecretName := secret.ObjectMeta.Name + "-backup"
+
+		secretBackupExists, err := SecretBackupExists(backupSecretName, TerraformStateNamespace, kubeCl, input)
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:65]:
```suggestion
		secretBackupExists, err := isSecretBackupExists(backupSecretName, TerraformStateNamespace, kubeCl, input)
```

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+		if secretBackupExists && err == nil {
+			input.LogEntry.Infof("Secret backup with name %s/%s already exists! Migration is not needed for Secret/%s/%s", TerraformStateNamespace, backupSecretName, TerraformStateNamespace, secret.ObjectMeta.Name)
+			continue
+		}
+		input.LogEntry.Infof("Secret backup with name %s/%s doesn't exist! Continue migration for Secret/%s/%s", TerraformStateNamespace, backupSecretName, TerraformStateNamespace, secret.ObjectMeta.Name)
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:70]: unnecessary

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+		terraformState := gjson.ParseBytes(terraformStateRaw)
+		openstackV2Resources := terraformState.Get("resources.#(type==\"openstack_blockstorage_volume_v2\")#")
+
+		if len(openstackV2Resources.Array()) == 0 {
+			input.LogEntry.Infof("No old resources found. Migration is not needed for Secret/%s/%s.", TerraformStateNamespace, secret.ObjectMeta.Name)
+			continue
+		}
+
+		input.LogEntry.Infof("Found old resources. Starting migration for Secret/%s/%s", TerraformStateNamespace, secret.ObjectMeta.Name)
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:88]: unnecessary

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+					terraformStateRaw, err = sjson.SetBytes(terraformStateRaw, fmt.Sprintf("resources.%d.instances.0.dependencies.%d", i, j), strings.ReplaceAll(terraformDependency.String(), OpenstackV2ResourceType, OpenstackV3ResourceType))
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		var newTerraformState bytes.Buffer
+		err = json.Indent(&newTerraformState, terraformStateRaw, "", "  ")
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:123]: why we need ident?

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
some nits

File: ee/candi/cloud-providers/openstack/layouts/standard-with-no-router/master-node/outputs.tf [hunk 8]
```diff
@@ -6,7 +6,7 @@ output "master_ip_address_for_ssh" {
 }
 
 output "node_internal_ip_address" {
-  value = module.master.node_internal_ip_address
+  value = module.master.node_networks[1].fixed_ip_v4
```
Comment (RomanenkoDenys) [inline ee/candi/cloud-providers/openstack/layouts/standard-with-no-router/master-node/outputs.tf:9]: it's bad solution, need to discuss

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go (hunk 1 above)
Comment (AleksZimin) [reply ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:39]: I've removed this line: `input.LogEntry.Infof("Starting Terraform migrate hook")`

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go (hunk 3 above)
Comment (AleksZimin) [reply ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:59]: `errors.IsNotFound(err)` doesn’t work for method List. List returns nil error if secrets with selected labels don’t exist.

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go (hunk 5 above)
Comment (AleksZimin) [reply ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:70]: I've removed this line: `input.LogEntry.Infof("Secret backup with name %s/%s doesn't exist! Continue migration for Secret/%s/%s", TerraformStateNamespace, backupSecretName, TerraformStateNamespace, secret.ObjectMeta.Name)
`

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go (hunk 6 above)
Comment (AleksZimin) [reply ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:88]: I've removed this line: `input.LogEntry.Infof("Found old resources. Starting migration for Secret/%s/%s", TerraformStateNamespace, secret.ObjectMeta.Name)`

**Review by z9r5 — DISMISSED** [review]

File: ee/candi/cloud-providers/openstack/openapi/cluster_configuration.yaml [hunk 9]
```diff
@@ -155,7 +161,7 @@ apiVersions:
                 items:
                   type: string
                 uniqueItems: true
-              additionalTags:
+              additionalTags: &instanceClassAdditionalTags
```
Comment (z9r5) [inline ee/candi/cloud-providers/openstack/openapi/cluster_configuration.yaml:164]:
```suggestion
              additionalTags: &instanceClassAdditionalTags
```

As I know, we agreed not to use YAML anchors.

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+		_, err = kubeCl.CoreV1().
+			Secrets(TerraformStateNamespace).
+			Update(context.TODO(), &secret, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+		input.LogEntry.Infof("Migration process completed for Secret/%s/%s.", TerraformStateNamespace, secret.ObjectMeta.Name)
+	}
+
+	input.LogEntry.Infof("Terraform migrate hook has been finished.")
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:134]: i think we should remove this log entry

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+	input.LogEntry.Debugf("Function isSecretBackupExists: err !=nil and errors.IsNotFound(err) no true. Return false and err")
+	return false, err
+}
+
+func BackupSecret(backupSecretName string, secret v1.Secret, namespace string, kubeCl k8s.Client, input *go_hook.HookInput) bool {
+	nodeName := secret.ObjectMeta.Labels["node.deckhouse.io/node-name"]
+	nodeGroup := secret.ObjectMeta.Labels["node.deckhouse.io/node-group"]
+	// nodeName := strings.TrimPrefix(secret.ObjectMeta.Name, "d8-node-terraform-state-")
+
+	input.LogEntry.Infof("Starting backup for Secret/%s/%s", namespace, secret.ObjectMeta.Name)
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:164]: i think we should remove this log entry

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+	_, err := kubeCl.CoreV1().
+		Secrets(namespace).
+		Create(context.TODO(), secretBackup, metav1.CreateOptions{})
+
+	if err != nil {
+		input.LogEntry.Warnf("An error occurred when creating secret backup. Backup aborted. Error: %s.", err)
+		return false
+	}
+	input.LogEntry.Infof("Secret backup completed successfully.")
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:188]: i think we should remove this log entry

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
less logs

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+		}
+
+		secret.Data[TerraformStateDataKey] = newTerraformState.Bytes()
+		_, err = kubeCl.CoreV1().
+			Secrets(TerraformStateNamespace).
+			Update(context.TODO(), &secret, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+		input.LogEntry.Infof("Migration process completed for Secret/%s/%s.", TerraformStateNamespace, secret.ObjectMeta.Name)
```
[the remark this answers is not in the document]
Comment (AleksZimin) [reply ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:131]: I've removed this line: `input.LogEntry.Infof("Migration process completed for Secret/%s/%s.", TerraformStateNamespace, secret.ObjectMeta.Name)`

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go (hunk 10 above)
Comment (AleksZimin) [reply ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:134]: I've removed this line: `input.LogEntry.Infof("Terraform migrate hook has been finished.")`

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go (hunk 11 above)
Comment (AleksZimin) [reply ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:164]: I've removed this line: `input.LogEntry.Infof("Starting backup for Secret/%s/%s", namespace, secret.ObjectMeta.Name)`

File: ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go (hunk 12 above)
Comment (AleksZimin) [reply ee/modules/030-cloud-provider-openstack/hooks/migrate/migrate_terraform_state.go:188]: I've removed this line: `input.LogEntry.Infof("Secret backup completed successfully.")`

**Review by konstantin-axenov — APPROVED** [review]

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/openstack'...

**Pull request merged**